runtime/src/main/java/org/corfudb/runtime/collections/ISMRMap.java

```diff
@@ -255,4 +265,87 @@ default void undoPutAll(ISMRMap<K,V> map, Map<K,V> undoRecord, Map<? extends K,
     @Accessor
     @Override
     Set<Entry<K, V>> entrySet();
+
+    @Override
+    default List<ISMREntry> coalesceUpdates(List<ISMREntry> updates, ISMREntryGenerator generator) {
```

Comment:
Just a suggestion to make this code more readable. Coalescing can happen in 3 scans only. In particular:

1. Expand: expand all putAll calls to only put calls)

2. Trim: While iterating the expanded list backwards, you just need to keep the latest key in the list (can be done while maintaining a map of keys seen so far). If you see a key that is already in your map, you can just ignore it.

3. Compress: By #1 and #2 you'll be left with a list of puts and removes only that are independent from each other. All the puts can be grouped in one putAll call and the removes can be left untouched.

Obviously, both methods have the same side effect, but I feel that the aforementioned simplifies the merge/resolution logic.